### ISSUE TITLE
Boolean unite() results in a deformed path

### ISSUE THREAD

**Issue by iconexperience** [paperjs/paper.js#762]
In some cases when calling the <code>unite()</code> function on two paths, the result is a slightly deformed path. Here is an example
```js
var p1 = new Path({segments:[[463.9780047854934, 242.45788478669098, 0, 0, 18.39490991498633, 18.763283770414063], [503.97992517441867, 317.201107498782, -7.581784931122968, -30.770871629359817, 0, 0], [445.72229285125354, 331.5554892722653, 0, 0, -5.274450958793182, -21.406496602422067], [421.13310212488653, 284.46162647720143, 10.14208864284501, 10.345192670683218, 0, 0]]});
var p2 =  new Path({segments:[[503.9799251744186, 317.2011074987819, 0, 0, 10.51261931443213, 42.66573931487159], [493.7273421855935, 475.64521896494347, 18.87625636417505, -62.72422687679028, 0, 0], [436.2726578144065, 458.35478103505653, 0, 0, 16.587797481872997, -55.11986871580717], [445.7222928512536, 331.5554892722654, 7.346578331020112, 29.81627951630253, 0, 0]]});
p1.closePath();
p2.closePath();
p1 = cleanUp(p1);
p2 = cleanUp(p2);
p = p1.unite(p2);
p.strokeColor = "red";
p1.strokeColor = "green";
p2.strokeColor = "green";
```
As you can see in this image, there is a clear difference between the united path and the two single paths.
![image](https://cloud.githubusercontent.com/assets/2122773/9539434/9f4250ca-4d51-11e5-9a37-a6ce5f3c17f8.png)
For some reason the point where the two original paths touch is discarded and another point is added, which then leads to the observed result.

**Comment by lehni** [comment]
This one too is scale related, which is good: I have something to compare it too. Will investigate:

```js
var p1 = new Path({segments:[[463.9780047854934, 242.45788478669098, 0, 0, 18.39490991498633, 18.763283770414063], [503.97992517441867, 317.201107498782, -7.581784931122968, -30.770871629359817, 0, 0], [445.72229285125354, 331.5554892722653, 0, 0, -5.274450958793182, -21.406496602422067], [421.13310212488653, 284.46162647720143, 10.14208864284501, 10.345192670683218, 0, 0]]});
var p2 =  new Path({segments:[[503.9799251744186, 317.2011074987819, 0, 0, 10.51261931443213, 42.66573931487159], [493.7273421855935, 475.64521896494347, 18.87625636417505, -62.72422687679028, 0, 0], [436.2726578144065, 458.35478103505653, 0, 0, 16.587797481872997, -55.11986871580717], [445.7222928512536, 331.5554892722654, 7.346578331020112, 29.81627951630253, 0, 0]]});
p1.closePath();
p2.closePath();
project.activeLayer.scale(2);
p1 = cleanUp(p1);
p2 = cleanUp(p2);
p = p1.unite(p2);
p.strokeColor = "red";
p1.strokeColor = "green";
p2.strokeColor = "green";

function cleanUp(path) {
    var newPath = path.unite();
    path.remove();
    return newPath;
}
```

**Comment by iconexperience** [comment]
I will try to investigate this in parallel, because I need to understand better how the boolean operations work and this is a good opportunity. But I suspect you will be faster.

**Comment by lehni** [comment]
This one boils down to an error in the fat-line clipping code...

```js
var curve1 = new Curve(463.9780047854934,242.45788478669098,482.3729147004797,261.22116855710505,496.3981402432957,286.4302358694222,503.97992517441867,317.201107498782);
var curve2 = new Curve(503.9799251744186,317.2011074987819,514.4925444888507,359.86684681365347,512.6035985497685,412.9209920881532,493.7273421855935,475.64521896494347);

curve1.getIntersections(curve2).forEach(function(inter) {
    console.log(inter, inter._overlap, inter.getIntersection()); 
});
```

**Comment by lehni** [comment]
While a correct result with a almost identical curves are simply:

```
{ point: { x: 503.97993, y: 317.20111 }, parameter: 1 } undefined { point: { x: 503.97993, y: 317.20111 }, parameter: 0 }
```

**Comment by lehni** [comment]
And yes, it took me quite a while to get familiar with it all.... By now I have a pretty solid intuition about such errors : )

**Comment by lehni** [comment]
This one's hard for me to debug though, as I have very little experience with the fat-line clipping code. You looked into it quite a bit a while back, @iconexperience ? Fancy tackling it?

**Comment by iconexperience** [comment]
I think that I have some understanding of the fat line clipping, so please assign this issue to me. I will try my best.

**Comment by lehni** [comment]
Cool thanks!

**Comment by lehni** [comment]
I think I need to add you to the team first. Are you happy to join? :)

**Comment by iconexperience** [comment]
I am not sure if I am ready yet. Can we have me assigned informally?

**Comment by lehni** [comment]
Yes, sure! : )

**Comment by lehni** [comment]
To explain the logged information: the left object is the intersection object, while the right is the linked intersection object on the other curve. As you can see, I get three intersections for the same location on the left, which is clearly no the case with these curves.

**Comment by iconexperience** [comment]
Yes, I understand. Here is the visualized result:

![image](https://cloud.githubusercontent.com/assets/2122773/9549416/6443c620-4da6-11e5-8c43-d4be99a74425.png)

The dots show the found intersections points on curve 2.

**Comment by lehni** [comment]
Yes. Bizarre init?

**Comment by lehni** [comment]
In comparison, here a pair of almost identical curves where the algorithm works as it should (new constructor notation requires changes to Curve constructor in `boolean-fix` code that allow construction directly from `toString()` output):

```js
var curve1 = new Curve({ point1: { x: 463.978, y: 242.45788 }, handle1: { x: 18.39491, y: 18.76328 }, handle2: { x: -7.58178, y: -30.77087 }, point2: { x: 503.97993, y: 317.20111 } });
var curve2 = new Curve({ point1: { x: 503.97993, y: 317.20111 }, handle1: { x: 10.51262, y: 42.66574 }, handle2: { x: 18.87626, y: -62.72423 }, point2: { x: 493.72734, y: 475.64522 } });

curve1.getIntersections(curve2).forEach(function(inter) {
    console.log(inter, inter.getIntersection()); 
});```

**Comment by lehni** [comment]
And this, along with some logging, leads to a solution: `tDiff` is getting extremely close to 0, leading to a new value for tDiff in the next recursion that is very far off, due to imprecision.

**Comment by iconexperience** [comment]
Excellent! I had already found the small value for tDiff, but did not see where to handle these cases.

After testing yourbugfix I can say that the deformations have completely disappeared. I have not been able to reproduce this issue even once.

**Comment by iconexperience** [comment]
@lehni You may want to consider changing the function <code>addCurveIntersections</code> a little bit here:
<pre>
tMinClip = clipConvexHull(top, bottom, dMin, dMax);
top.reverse();
bottom.reverse();
tMaxClip = clipConvexHull(top, bottom, dMin, dMax);
if (tMinClip == null || tMaxClip == null)
	return;
</pre>

In my test in about 50% of all calls to <code>addCurveIntersections</code>, <code>clipConvexHull(top, bottom, dMin, dMax);</code> returned <code>null</code> for <code>tMinClip</code>. In all these cases we make an unnecessary second call <code>clipConvexHull(...)</code>. If you check for <code>tMinClip == null</code> earlier, you could avoid all these calls. Here is the changed code part:

<pre>
tMinClip = clipConvexHull(top, bottom, dMin, dMax);
if (tMinClip == null) return;
top.reverse();
bottom.reverse();
tMaxClip = clipConvexHull(top, bottom, dMin, dMax);
if (tMaxClip == null) return;
</pre>

I did not think this was worth mentioning before, but since it happens in about 50% of all cases, the change may actually give a little performance improvement, even though it increases code size a little bit.

**Comment by iconexperience** [comment]
And while you are at this, the comment could be corrected from

<pre>
// No intersections if one of the tvalues are null or 'undefined'
</pre>

to

<pre>
// No intersections if one of the tvalues are null
</pre>

because undefined will never be returned by <code>clipConvexHull(hullTop, hullBottom, dMin, dMax) </code>

**Comment by lehni** [comment]
@iconexperience yes, you're right on both accounts. Thanks!

**Comment by lehni** [comment]
@iconexperience are you saying that other than the unresolved self-intersections, there is no new edge case that you can produce or observe? That's very good news! Who would have thought we'd get here this quickly! 🎉

**Comment by iconexperience** [comment]
@lehni 
Well, currently I cannot deliver any problematic edge cases, but I still do not have much time for testing. 

Anyway, these cases are not a problem anymore. Check this test case:

<pre>
createRandomCurve: function(size) {
        var x1 = Math.random() * size + size;
        var y1 = Math.random() * size + size;
        var x2 = Math.random() * size + size;
        var y2 = Math.random() * size + size;
        var hx1 = (Math.random() - .5) * 2 * size;
        var hy1 = (Math.random() - .5) * 2 * size;
        var hx2 = (Math.random() - .5) * 2 * size;
        var hy2 = (Math.random() - .5) * 2 * size;
        return new paper.Curve(new paper.Point(x1, y1), new paper.Point(hx1, hy1), new paper.Point(hx2, hy2), new paper.Point(x2, y2));
    },

 var cycles = 1000000;
    var failCount = 0;
    for (var i = 0; i < cycles; i++) {
        crv1 = createRandomCurve(300);
        crv2 = crv1.divide(0.5, true);
        ints = crv1.getIntersections(crv2);
        for (var j = 0; j < ints.length; j++) {
            var loc = ints[j];
            if (loc.curve.getPointAt(loc.parameter, true).getDistance(loc._curve2.getPointAt(loc._parameter2, true)) > 0.01) {
              failCount++;
            }
        }
    }
    console.log("Problems in " + (100 * failCount / cycles) + "% of cases");
</pre>

Before your bugfix 0.25% of all cases resulted in wrong intersections, now we are at 0.00% (and I ran 10,000,000 cycles already). This is great!

**Comment by iconexperience** [comment]
@lehni One other question: Do you think the code for determining if a path is clockwise is bullet proof? When looking at it I get a strange feeling that treating the handles just like points on a polygon may give wrong results in some cases. Do you think I should investigate this, or are you confident about this part of the code?

**Comment by lehni** [comment]
@iconexperience found a way to address your code comments without the code growing in size : )

**Comment by lehni** [comment]
@iconexperience regarding clockwise, it's based on these two posts. Not sure how trustworthy the first one is:

http://stackoverflow.com/questions/3308741/how-to-find-orientation-of-a-bezier-curve
http://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-points-are-in-clockwise-order

**Comment by iconexperience** [comment]
@lehni I believe the showlace algorithm of the second post just works, otherwise a smart person would have found out. But the first one is a bit shaky. I will better take a closer look. Is it required that the algorithm works for self intersecting paths?

**Comment by lehni** [comment]
@iconexperience I don't think it's a requirement for that... it's only used in `reorient()`, to fix orientations of children of compound paths...

**Comment by iconexperience** [comment]
@lehni it is not needed for boolean operations?

**Comment by lehni** [comment]
It is used once in the beginning, through `reorient()`, but it hasn't been the source of troubles so far. We haven't encountered a path yet where the result of the algorithm would have been found to be wrong.
